lib/Service/TokenService.php

```diff
@@ -0,0 +1,276 @@
+	 * @throws MultipleObjectsReturnedException
+	 * @throws TokenExchangeFailedException
+	 * @throws \JsonException
+	 */
+	public function getExchangedToken(string $targetAudience): Token {
+		$loginToken = $this->getToken();
+		if ($loginToken === null) {
+			$this->logger->debug('[TokenService] Failed to exchange token, no login token found in the session');
+			throw new TokenExchangeFailedException('Failed to exchange token, no login token found in the session');
+		}
+		if ($loginToken->isExpired()) {
+			$this->logger->debug('[TokenService] Failed to exchange token, the login token is expired');
+			throw new TokenExchangeFailedException('Failed to exchange token, the login token is expired');
+		}
+		$oidcProvider = $this->providerMapper->getProvider($loginToken->getProviderId());
+		$discovery = $this->discoveryService->obtainDiscovery($oidcProvider);
+
+		try {
+			$clientSecret = $oidcProvider->getClientSecret();
+			if ($clientSecret !== '') {
+				try {
+					$clientSecret = $this->crypto->decrypt($clientSecret);
+				} catch (\Exception $e) {
+					$this->logger->error('[TokenService] Token Exchange: Failed to decrypt oidc client secret');
+				}
+			}
+			$this->logger->debug('[TokenService] Exchanging the token: ' . $discovery['token_endpoint']);
+			// more in https://www.keycloak.org/securing-apps/token-exchange
+			$result = $this->client->post(
+				$discovery['token_endpoint'],
+				[
+					'body' => [
+						'client_id' => $oidcProvider->getClientId(),
+						'client_secret' => $clientSecret,
+						'grant_type' => 'urn:ietf:params:oauth:grant-type:token-exchange',
+						'subject_token' => $loginToken->getAccessToken(),
+						'subject_token_type' => 'urn:ietf:params:oauth:token-type:access_token',
+						// can also be
+						// urn:ietf:params:oauth:token-type:access_token
+						// or urn:ietf:params:oauth:token-type:id_token
+						// this one will get us an access token and refresh token within the response
+						'requested_token_type' => 'urn:ietf:params:oauth:token-type:refresh_token',
+						'audience' => $targetAudience,
+					],
+				]
+			);
+			$this->logger->debug('[TokenService] Token exchange request params', [
+				'client_id' => $oidcProvider->getClientId(),
+				// 'client_secret' => $clientSecret,
+				'grant_type' => 'urn:ietf:params:oauth:grant-type:token-exchange',
+				// 'subject_token' => $loginToken->getAccessToken(),
+				'subject_token_type' => 'urn:ietf:params:oauth:token-type:access_token',
+				'requested_token_type' => 'urn:ietf:params:oauth:token-type:refresh_token',
+				'audience' => $targetAudience,
+			]);
+			$body = $result->getBody();
+			$bodyArray = json_decode(trim($body), true, 512, JSON_THROW_ON_ERROR);
+			$this->logger->debug('[TokenService] Token exchange success: "' . trim($body) . '"');
+			$tokenData = array_merge(
+				$bodyArray,
+				['provider_id' => $loginToken->getProviderId()],
+			);
+			return new Token($tokenData);
+		} catch (\Exception|\Throwable $e) {
```

Comment:
Currently, if the audience is invalid, the code throws an error:
`Failed to exchange token, error in the exchange request.`

It would be better if it throws a more specific error for an invalid audience, as this would make it clearer and more helpful, especially for displaying in the UI.

On API, when the audience is invalid, we get status code `400` and the response:

```json
{
  "error": "invalid_client",
  "error_description": "Audience not found"
}
```

Reply:
Good point. The `TokenExchangeFailedException` class now has 2 additional methods:
```
public function getError(): ?string;
public function getErrorDescription(): ?string;
```
They both return a string only if the exchange request fails and returns a JSON object with the `error` and `error_description` keys.